[WIP] liblokinet changes

majestrate opened the pull request:
placeholder for dan's branch for liblokinet used for code review.

tewinget committed:
change llarp/quic log statements
- add new format with category
- add set_data_dir method to liblokinet

tewinget committed:
bump ngtcp2 version, working but incomplete

bump ngtcp2 dep to the latest tagged version.  The changes to get
liblokinet/quic back to "working" (insofar as it was before) are
complete in this commit, but there appear to be further changes in
ngtcp2 apart from simple API changes which could merit reworking how we
use it somewhat.

tewinget committed:
ios liblokinet changes
- fix some ngtcp2 usage and conn issue detection
- clients now get an ERR_PROTO packet when something goes wrong and the
server is unhappy.
- ngtcp2_accept changed between our last update of it and current, this
fixes its usage.
- with a minor change to the C API, liblokinet userland will now be able
to receive notification when a connection is (un)successfully opened,
and when it closes.  TODO: supply a reason for the close, if we decide
it should

majestrate committed:
recuce number of jobs pushed through omq proxy thread.

when we queue cpu heavy work in lokinet to worker threads we make 1
job per function call. we call a lot of jobs so this coleases the jobs
into 1 job that we push off at the end of the event loop cycle,
reducing the number of jobs going across the omq proxy thread, in
theory reducing cpu usage.

(cherry picked from commit bbbb07d01d8dd99ce5cdda9701107bf1187e1767)

dr7ana committed:
added required callback (version_negotiation)

dr7ana committed:
updated cmake to have lokinet-embedded-api build deps
- pkt tx timer update + max pkt count
- added pkt tx time updates to message handling logic
- setting destlen param to at least max_tx_UDP_Payload size

majestrate commented on llarp/quic/connection.cpp:
```diff
@@ -897,35 +897,42 @@ namespace llarp::quic
       assert(consumed <= 0);
 
-      switch (nwrite)
+      if (nwrite == -240)
```
there isn't really a good reason to use switch case for anything not an `enum class`. the compiler is smart enough to optimize it.

dr7ana replied:
agreed. reverted to a if/else block after this for better control over breaking/continuing

majestrate commented on llarp/lokinet_shared.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <lokinet.h>
+#include <filesystem>
```
this should be `llarp/util/fs.hpp` as it contains compat shims for macos and such.

dr7ana committed:
debugging

dr7ana committed:
ngtcp2 timing statements

dr7ana committed:
timing prints for testnet

dr7ana committed:
debugging, squash

dr7ana committed:
there do you see tom??

dr7ana committed:
save

dr7ana committed:
is this working?

dr7ana committed:
returned for loop

dr7ana committed:
breaking testnet today i suppose

dr7ana committed:
breaking testnet today i suppose pt 2: the breakening

dr7ana committed:
breaking testnet today i suppose pt 2: the breakening

dr7ana committed:
breaking testnet today i suppose pt 3: the re-breakening

dr7ana committed:
breaking testnet today i suppose pt 3: the re-breakening

dr7ana committed:
breaking testnet today i suppose pt 3: the re-breakening

dr7ana committed:
null crypto debugging

dr7ana committed:
edge case for write close packet error on testnet

dr7ana committed:
edge case for write close packet on testnet

dr7ana committed:
null crypto thangs

dr7ana committed:
why is client transport params not being understood by server

dr7ana committed:
lfg

jagerman commented on llarp/quic/endpoint.hpp:
```diff
@@ -111,7 +111,7 @@ namespace llarp::quic
     // Called to handle an incoming packet
     void
-    handle_packet(const Packet& p);
+    handle_packet(Packet p);
```
This would be better as an rvalue reference `Packet&& p` so that we are still passing it by reference rather than needing to copy it (there is internal ngtcp2 C stuff in here that likely isn't moveable, so taking it by value will involve at least *some* copying).  Whereas if you take it as a `Packet&&` you *have* to move it when calling, but we still take it as a reference so don't have to copy anything.

majestrate replied:
hah, i knew it

dr7ana committed:
addressing fix plus linting

dr7ana committed:
Revert "addressing fix plus linting"

This reverts commit 550486d216935389d0bb0f25b73a4d0b8fd11d36.

dr7ana committed:
Revert "lfg"

This reverts commit 6d28d0d78c35bf36b69cf53fa9003bc6158fe406.

dr7ana commented:
### Addressing Fix

1. Add an `std::optional<std::variant<llarp::service::Address, llarp::RouterID>>` member to `llarp::quic::Address`
2. Set it in `.remote` when we create initialize the `llarp::quic::Path` for `Connection` (I think `GetEndpointWithConvoTag` will give us what we want).
3. Hard-code `.remote` address to be `::1` (IPv6 localhost).
4. Chase through the code that constructs a `Connection` in `llarp::quic::Server` and `llarp::quic::Client` to stop using a `ConvoTag` and instead set the address variant.
5. Remove the `ConvoTag` stuff from `llarp::quic::Endpoint`
6. Where we call `SendToOrQueue` in `llarp::quic::Endpoint` we now pass the variant (instead of, currently, the `ConvoTag` via implicit conversion)

dr7ana committed:
massive code change to remove ConvoTag from llarp/quic

dr7ana committed:
debug statements

dr7ana committed:
Addressing fix

client->server:

  pseudo-port = 0
  Path =
      local = [::1]:0
      remote = [::1]:12345 <-- remote-port

server receives:

  pseudo-port = 0
  Path =
      local = [::1]:0 <-- port is pseudo-port, set in quic/endpoint.cpp
      remote = [::1]:0 <-- port is pseudo-port, set in quic/tunnel.cpp

server sends:

  pseudo-port = 0
  Path =
      local = [::1]:0 <-- is what it saved above
      remote = [::1]:0 <-- is what it saved above

client receives:

  pseudo-port = 0
  Path =
      local = [::1]:12345 <-- port is remote-port, set in quic/endpoint.cpp
      remote = [::1]:12345 <-- port is remote-port, set in quic/tunnel.cpp

  ^^^^^^ This does not match the stored Path, so dropped.

dr7ana committed:
missing param in log statement breaks build

dr7ana committed:
removed print statement causing xcode crash

dr7ana committed:
debugging

dr7ana committed:
port juggling

dr7ana committed:
pseudo?

dr7ana committed:
port

dr7ana committed:
port

dr7ana committed:
endianness

dr7ana committed:
idle_expiry

dr7ana committed:
that was a bad idea

dr7ana committed:
removed idle expiry check from endpoint::check_timeouts, modified Connection::close_connection to handle idle expired connections without calling write_connection_close

dr7ana committed:
added logic to handle handshake failure

dr7ana committed:
changed error handling to use int to properly detect codes

dr7ana committed:
formatting

dr7ana committed:
exposed callbacks for TunnelManager::open and ::close s.t. session can pass functions that handle success/failure on open and close

dr7ana committed:
exploring reduced build size, unlinking libunbound from lokinet-embedded

majestrate commented on llarp/CMakeLists.txt:
we will need to make this a compile time flag so we can not compile in the libunbound parts in our dns subsystem. this wont really do anything right now.

however, we can also remove libssl and libexpat as a result which will make the static targets smaller by a lot probably

dr7ana replied:
sorry jeff just saw this! sounds good to me, will implement if it works in reducing the size as we think it will

majestrate replied:
There are a few things we need to do to make this compile time configurable:

- move libunbound resolver to its own compilation unit. 
- only add that new compiliation unit when libunbound is enabled
- in header for new compilation unit declare a single function that will create a shared_ptr of resolver_base for the libunbound resolver, add nop implementation of function in dns/server.cpp when libunvound is disabled on compile time.

majestrate replied:
There are a few things we need to do to make this compile time configurable:

- move libunbound resolver to its own compilation unit. 
- only add that new compiliation unit when libunbound is enabled
- in header for new compilation unit declare a single function that will create a shared_ptr of resolver_base for the libunbound resolver, add nop implementation of function in dns/server.cpp when libunvound is disabled on compile time.

dr7ana committed:
cleaning up repo
- removed fprintf debug statements used for xcode debugging that were failing CI
- removed unused variables also failing CI
- deb stable i386 (and others) broke with logging of variants

dr7ana committed:
error handling fix
- handshake timeouts was incorrect: must call ngtcp2_write_connection_close if NGTCP2_ERR_HANDSHAKE_TIMEOUT is returned
- packet handling fix: do not close stream on congestion
- bumped clang-format version to latest (16)
- format-version.sh checks for version 14 or greater now, removed old comment block

majestrate commented on llarp/endpoint_base.hpp:
```diff
@@ -122,7 +122,7 @@ namespace llarp
     EnsurePathTo(
         AddressVariant_t addr,
-        std::function<void(std::optional<service::ConvoTag>)> hook,
+        std::function<void(std::optional<std::variant<service::Address, RouterID>>)> hook,
```
`std::function<void(bool)>`

majestrate commented on llarp/net/ip_address.hpp:
```diff
@@ -24,7 +24,7 @@ namespace llarp
     /// Empty constructor.
     IpAddress() = default;
     /// move construtor
-    IpAddress(IpAddress &&) = default;
```
bin this entire file if you can

Goop-arsh replied:
https://github.com/oxen-io/lokinet/releases/tag/v0.9.11

tewinget committed:
fix bad_variant_access resulting from ConvoTag->AddressVariant change

We changed the quic code to refer to its remote endpoint using an Address Variant
and this resulted in a crash when sending a connection close packet.  This code was
incorrectly supplying an uninitialized value for ConvoTag before but this was silently
ignored; after switching to a variant it became a crash.

dr7ana committed:
refactoring liblokinet export fxns

dr7ana committed:
tom/jason-fix in lokinet_shared.cpp for lokinet_outbound_tcp

dr7ana committed:
linting

dr7ana committed:
CI is breaking on Debian stable builds due to oxen-logger formatting std::chrono

tewinget committed:
change how tcp connection closes are handled

this still feels incorrect, but more correct than before

tewinget committed:
ack to data we no longer have is not an error

tewinget committed:
fix segfault in lokinet_shared

tewinget committed:
avoid potential deadlock

tewinget committed:
few more unlikely but potential deadlock cases

locking in lokinet_shared needs revisited in general.  In addition, user-supplied
callbacks we call from the main thread currently should be called on a separate
thread instead, both to keep the main loop fast and to avoid complex locking issues.

dr7ana committed:
struct initialization

dr7ana committed:
re-inserting lokinet_outbound_tcp fix after rebasing with tomfix

majestrate committed:
Redo wire protocol race condition fix.

In commit a76acd49560f9845c2ad9a839364f2834afc7756 we attempted to fix
issues on service nodes that related to spamming service nodes with
traffc, but this caused a regression where clients could not connect
to the network.

However, we also did something idiotic with the protocol handshake.
The handshake logic was modified and caused client connections to
fail. This was due to a change in how we decided what a timed out
session is. We redo the close bug and revert the change to the
protocol handshake.

(cherry picked from commit b0b6e2b198ba32e8c5641e99494d2874b2243b8f)

dr7ana committed:
linting

dr7ana committed:
Merge pull request #1 from dr7ana/liblokinet-cherrypick

Merging cherry-picks back to testnet branch:
-  #2164
-  #2134

Goop-arsh commented on .drone.jsonnet:
```diff
@@ -315,6 +315,34 @@ local mac_builder(name,
+      environment: { SSH_KEY: { from_secret: 'SSH_KEY' } },
+      commands: [
+        'echo "Building on ${DRONE_STAGE_MACHINE}"',
+        'export CMAKE_BUILD_PARALLEL_LEVEL=6',
```
```suggestion
        'export CMAKE_BUILD_PARALLEL_LEVEL=6',
```
